test/unit/picolibrary/output_stream/main.cc

```diff
@@ -0,0 +1,864 @@
+    EXPECT_EQ( result.error(), error );
+
+    EXPECT_FALSE( stream.end_of_file_reached() );
+    EXPECT_FALSE( stream.io_error_present() );
+    EXPECT_TRUE( stream.fatal_error_present() );
+}
+
+/**
+ * \brief Verify picolibrary::Output_Stream::put( char ) works properly.
+ */
+TEST( putChar, worksProperly )
+{
+    auto stream = Mock_Output_Stream{};
+
+    auto const character = random<char>();
+
+    EXPECT_CALL( stream.buffer(), put( SafeMatcherCast<char>( Eq( character ) ) ) )
+        .WillOnce( Return( Result<Void, Error_Code>{} ) );
+
+    EXPECT_FALSE( stream.put( character ).is_error() );
+
+    EXPECT_TRUE( stream.is_nominal() );
+}
+
+/**
+ * \brief Verify picolibrary::Output_Stream::put( char const *, char const * ) properly
+ *        handles the presence of an I/O error and/or a fatal error.
+ */
+TEST( putCharBlock, errorPresent )
+{
+    auto stream = Mock_Output_Stream{};
+
+    stream.report_random_error();
+
+    auto const io_error_present    = stream.io_error_present();
+    auto const fatal_error_present = stream.fatal_error_present();
+
+    EXPECT_CALL( stream.buffer(), put( A<std::string>() ) ).Times( 0 );
+
+    auto const string = random_container<std::string>();
+    auto const result = stream.put( &*string.begin(), &*string.end() );
+
+    ASSERT_TRUE( result.is_error() );
+    EXPECT_EQ( result.error(), Generic_Error::IO_STREAM_DEGRADED );
+
+    EXPECT_FALSE( stream.end_of_file_reached() );
+    EXPECT_EQ( stream.io_error_present(), io_error_present );
+    EXPECT_EQ( stream.fatal_error_present(), fatal_error_present );
+}
+
+/**
+ * \brief Verify picolibrary::Output_Stream::put( char const *, char const * ) properly
+ *        handles a put error.
+ */
+TEST( putCharBlock, putError )
+{
+    auto stream = Mock_Output_Stream{};
+
+    auto const error = random<Mock_Error>();
+
+    EXPECT_CALL( stream.buffer(), put( A<std::string>() ) ).WillOnce( Return( error ) );
+
+    auto const string = random_container<std::string>();
+    auto const result = stream.put( &*string.begin(), &*string.end() );
+
+    ASSERT_TRUE( result.is_error() );
+    EXPECT_EQ( result.error(), error );
+
+    EXPECT_FALSE( stream.end_of_file_reached() );
+    EXPECT_FALSE( stream.io_error_present() );
+    EXPECT_TRUE( stream.fatal_error_present() );
+}
+
+/**
+ * \brief Verify picolibrary::Output_Stream::put( char const *, char const * ) works
+ *        properly.
+ */
+TEST( putCharBlock, worksProperly )
+{
+    auto stream = Mock_Output_Stream{};
+
+    auto const string = random_container<std::string>();
+
+    EXPECT_CALL( stream.buffer(), put( string ) ).WillOnce( Return( Result<Void, Error_Code>{} ) );
+
+    EXPECT_FALSE( stream.put( &*string.begin(), &*string.end() ).is_error() );
+
+    EXPECT_TRUE( stream.is_nominal() );
+}
+
+/**
+ * \brief Verify picolibrary::Output_Stream::put( char const * ) properly handles the
+ *        presence of an I/O error and/or a fatal error.
+ */
+TEST( putNullTerminatedString, errorPresent )
```

Comment:
This test (and its name and description) will need to be updated when error handling is revised.